field/TinyLang/Field/Generator.hs

```diff
@@ -269,6 +201,51 @@ arbitraryUnOpRing = elements [Neg]
 arbitraryBinOpRing :: MonadGen m => m (BinOp f (AField f) (AField f) (AField f))
 arbitraryBinOpRing = elements [Add, Sub, Mul]
 
+-- | Generate an expression of a particular type from a collection of variables
+-- with the number of nodes (approximately) bounded by 'size'.
+boundedArbitraryExpr
+    :: forall f a m. (Field f, Arbitrary f, KnownUni f a, MonadGen m, MonadSupply m)
+    => Vars f -> Int -> m (Expr f a)
+boundedArbitraryExpr _    size | size <= 1 = EVal <$> arbitraryM
+boundedArbitraryExpr vars size             = frequency $ everything ++ comparisons where
+    -- The most general generator.
+    everything =
+        [ (1, EVal <$> arbitraryM)
+        , (2, EVar <$> chooseUniVar vars)
+        , (2, do
+                let size' = size `Prelude.div` 3
```

Comment:
I'm sure I tried to do this myself and failed.  Maybe my mistake was to try to define variables containing size/2 and size/3 globally, so they'd apply to all cases.